### ISSUE TITLE
Add Scope for current route

### ISSUE THREAD

**Issue by thigg** [vaadin/cdi#343]
**Context:**
```
@SomeScope Controller{}
@Route class A{@Inject Controller c;}
@Route class B{@Inject Controller c;}
```
When I navigate from A to B, there should be a new Context.
But all subcomponents from A or B should receive the same controller.

Afaik currently I need a RouteScopeOwner annotation which chooses some route as scope.

**Non working solutions:**
As far as I understand the current capabilities, i could add a RouterLayout and let A and B use that RouterLayout. This does not solve the problem afaik, because when I navigate from A to B, the same context is reused.

A solution would be, to add multiple @RouteScopeOwner annotations to the controller and reference either A or B as RoutescopeOwner in A or B.
This would not work anymore if I add a component c to A and B, because it would have to figure out first, in which Route it is currently used.

**Proposed Solution**
This leads me to the conclusion, that a new scope would be useful.
It is very similar to the RouteScope but does not need a RouteScopeOwner annotation, because it always uses the current leaf as the owner.
See a sketch of the usage of the scope below:

```
@TheNewProposedScope
class Controller{
}
@TheNewProposedScope
class MyComponent extends Div{
@Inject Controller controller;
}

@Route 
class A extends Div{
@Inject public A(Controller controller, MyComponent comp){
  add(comp);
  controller.doSmth();
}
}

@Route 
class B extends Div{
@Inject public B(Controller controller, MyComponent comp){
  add(comp);
  controller.doSmth();
}
}
```

When navigating from A to B, there should be a new Controller and MyComponent.

I hope I made the usecase clear. Tell me, if I need to elaborate.
Also, of course, if either the usecase is a bad Idea in general, or it is already archivable.

I also had problems coming up with a good name for the scope/ the title. Feel free to edit.

**Comment by roeltje25** [comment]
I understand what you mean. Actually I would suggest that @RouteScoped annotation could be made to work like this. At this moment a @RouteScopeOwner is obligatory unless it is a Component with @Route annotation. In this case @RouteSCopeOWner is the component itself.

It seems to me a valuable addition if @RouteScoped beans (some controller e.g.) should have the "current route component"  as default RouteScopeOwner.

**Comment by thigg** [comment]
Nice, that actually seems to do the job (only a quick test, will test more). Can I help somehow to get that released in the near future?

**Comment by roeltje25** [comment]
It was just a quick jot. Could not even test it myself because of time
constaints. But if I understand your proposal correctly, I would like to
see such a feature implemented, that's why I jotted down a possible
solution.

Roel

--------------------------------------------
Roel Meeuws
Email: r.j.meeuws@gmail.com
Mob. phone: +31 (0)6 10 82 44 01
--------------------------------------------

**Comment by thigg** [comment]
[Backported to java8](https://github.com/vaadin/cdi/files/4770130/java8_port.patch.txt)
As I said, just let me know if I can help somehow.

**Comment by roeltje25** [comment]
Great apparently I had my idea set to Java 11. But does it help in your use
case?

Op vr 12 jun. 2020 12:17 schreef thigg <notifications@github.com>:

**Comment by thigg** [comment]
Yep, looks like it does exactly what I need.

**Comment by roeltje25** [comment]
Point is only that you should try to convince someone who has edit
permissions, I am just a helpful dev who liked the idea.

Roel

--------------------------------------------
Roel Meeuws
Email: r.j.meeuws@gmail.com
Mob. phone: +31 (0)6 10 82 44 01
--------------------------------------------

**Comment by thigg** [comment]
Currently proposed solution doesn't always get the right context. This is because `UI.getInternals().getActiveRouteChain()` does return the old route during the new route classes are constructed.

**Comment by roeltje25** [comment]
Hey Thigg,

Do you have a test case demonstrating the issue?

**Comment by thigg** [comment]
Could create something for you, but its quite easy:
Just create a route with constructor injection and print `UI.getInternals().getActiveRouteChain()`.
That will correspond to the context the route scoped injections are created in.

If the injected class is proxied, i guess it also changes during the lifetime of the class, when routing is finished.
That way the proxy is used on different objects before and after navigation.

**Comment by thigg** [comment]
Okay, i have a very dirty hack to fix this. I couldnt find a way to pass the current route down the chain, so I did this:

```
diff --git a/vaadin-cdi/src/main/java/com/vaadin/cdi/AbstractCdiInstantiator.java b/vaadin-cdi/src/main/java/com/vaadin/cdi/AbstractCdiInstantiator.java
index 488d78c..77539a4 100644
--- a/vaadin-cdi/src/main/java/com/vaadin/cdi/AbstractCdiInstantiator.java
+++ b/vaadin-cdi/src/main/java/com/vaadin/cdi/AbstractCdiInstantiator.java
@@ -19,7 +19,10 @@ import javax.enterprise.inject.spi.BeanManager;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.stream.Stream;
 
+import com.vaadin.cdi.context.RouteScopedContext;
+import com.vaadin.flow.component.HasElement;
 import com.vaadin.flow.internal.UsageStatistics;
+import com.vaadin.flow.router.NavigationEvent;
 import org.apache.deltaspike.core.api.provider.BeanProvider;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -50,6 +53,14 @@ abstract public class AbstractCdiInstantiator implements Instantiator {
                 && getServiceClass().isAssignableFrom(service.getClass());
     }
 
+    @Override
+    public <T extends HasElement> T createRouteTarget(Class<T> routeTargetType, NavigationEvent event) {
+        RouteScopedContext.currentRoute.set(routeTargetType);
+        final T result = this.getOrCreate(routeTargetType);
+        RouteScopedContext.currentRoute.set(null);
+        return result;
+    }
+
     @Override
     public <T> T getOrCreate(Class<T> type) {
         return new BeanLookup<>(getBeanManager(), type)
diff --git a/vaadin-cdi/src/main/java/com/vaadin/cdi/context/RouteScopedContext.java b/vaadin-cdi/src/main/java/com/vaadin/cdi/context/RouteScopedContext.java
index d9b489d..62f08ff 100644
--- a/vaadin-cdi/src/main/java/com/vaadin/cdi/context/RouteScopedContext.java
+++ b/vaadin-cdi/src/main/java/com/vaadin/cdi/context/RouteScopedContext.java
@@ -75,6 +75,8 @@ public class RouteScopedContext extends AbstractContext {
     private Supplier<Boolean> isUIContextActive;
     private BeanManager beanManager;
 
+    public static ThreadLocal<Class> currentRoute = new ThreadLocal<>();
+
     public RouteScopedContext(BeanManager beanManager) {
         super(beanManager);
     }
@@ -122,10 +124,8 @@ public class RouteScopedContext extends AbstractContext {
                 .map(annotation -> (Class) (((RouteScopeOwner) annotation).value()));
         Stream<Class> byBeanItself = Stream.of((Class) bean.getBeanClass())
                 .filter(b -> b.isAnnotationPresent(Route.class));
-        Stream<Class> byCurrentRoute = Optional.ofNullable(UI.getCurrent())
-                .map(ui -> ui.getInternals().getActiveRouterTargetsChain())
-                .map(s -> s.stream()
-                        .map(v->(Class)v.getClass()))
+        Stream<Class> byCurrentRoute = Optional.ofNullable(currentRoute.get())
+                .map(Stream::of)
                 .orElse(Stream.empty());
         return Stream.of(byAnnotation, byBeanItself, byCurrentRoute)
                 .map(Stream::findFirst)

```

**Comment by roeltje25** [comment]
This will also not work. remember that the a route may be UIScoped so when entering the location again, createRouteTarget is not invoked and your fix will not set the ThreadLocal again.

**Comment by mshabarov** [comment]
@thigg could you please explain what do you mean under 'Context':
> When I navigate from A to B, there should be a new Context.

This is not clear for me, why is it not working:
>As far as I understand the current capabilities, i could add a RouterLayout and let A and B use that RouterLayout. This does not solve the problem afaik, because when I navigate from A to B, the same context is reused.

Here is a project example  based on CDI skeleton starter, showing how to make `Controller` bean be limited to a single route component, hierarchy of route components or to UI:

[current-route-scope.zip](https://github.com/vaadin/cdi/files/6427191/current-route-scope.zip)

1. Download the project 
2. `mvn clean package tomee:run` and go to http://localhost:8080/a
3. Click the button to navigate between routes A and B and notice that the controller bean returns different UIDs (it's being re-created between navigations).
4. Uncomment `@NormalUIScoped` on top of `ControllerImpl` and see that now the controller bean is being preserved within entire UI (browser tab).
5. Comment out `@NormalUIScoped` and uncomment `@NormalRouteScoped` and `@RouteScopeOwner(ParentView.class)` on top of `ControllerImpl`. Also uncomment `@RouteScopeOwner(ParentView.class)` in the A and B classes. See that the controller bean is also preserved within `ParentView` router layout navigation.

Does this help?

If you meant some other use case, please elaborate and provide a code example showing your case.

**Comment by thigg** [comment]
I am not fully into the topic anymore, but I hope I can explain the problem correctly:
(cant test it right now)

- Have a view A with a subcomponent A1 and a view B. 
- A and B have the same parentroute ABParent.
-  A, B and A1 inject a controller GUIDController, which simply holds a GUID.
- When I go on route A, the guid of the injected GUIDController into A and A1 should be the same
- When I go on route B, the guid should change

If I remember the problem correctly, it was required to set a RouteScopeOwner. I could set the RouteScopeOwner to ABParent, but then the guid wouldnt change between A and B. If I set nothing, A and A1 get a different controller. If I set @NormalUIScoped, A, A1 and B have the same controller.

**Comment by mshabarov** [comment]
The following code should reflect your scenario:

```java
@Route("")
@RoutePrefix("parent")
public class ABParent extends Div implements RouterLayout {
    
    @Route(value = "a", layout = ABParent.class)
    public static class A extends Div {
        @Inject
        public A(@RouteScopeOwner(A.class) GUIDController controller,
                A1 a1) {
            add(new VerticalLayout(
                    new Span("A Controller = " + controller.getUid()),
                    a1,
                    new RouterLink("Navigate to B", B.class)));
        }
    }
    
    @Route(value = "b", layout = ABParent.class)
    public static class B extends Div {
        @Inject
        public B(@RouteScopeOwner(A.class) GUIDController controller) {
            add(new VerticalLayout(
                    new Span("B Controller = " + controller.getUid()),
                    new RouterLink("Navigate to A", A.class)));
        }
    }

    public static class A1 extends Div {
        @Inject
        public A1(@RouteScopeOwner(A.class) GUIDController controller) {
            add("A1 Controller = " + controller.getUid());
        }
    }

    @RouteScopeOwner(A.class)
    @NormalRouteScoped
    public static class GUIDController {
        private String uid = UUID.randomUUID().toString();
        public String getUid() {
            return uid;
        }
    }

}
```

It works for A and A1: they have the same instance of controller bean. But when I navigate from A to B, B uses also the same bean, but it is supposed to be a new one, because the old one is now out-of-scope.
Then, if i navigate back to A, a new bean re-created.
This looks like a bug and should be investigated.

My example shown above wouldn't have any sense, if it is extended by adding one more route component C with controller bean injected: `@RouteScopeOwner` should have the same owner for injected beans in all places. If I set A as a owner everywhere, it would mean the controller instances should be different for C and C1 (because they are out-of-scope of route A), which is not desired.

Thus, having a 'generic' capabilities would indeed require a new annotation to preserve beans within A, B and C, but re-create them after navigating between A, B and C. Using an existing `@NormalRouteScoped` annotation for this is not desired, see https://github.com/vaadin/spring/issues/785. So, it can be something like `@ParentNormalRouteScoped` to emphasise that the component having this bean is considered as a route scope owner.

To recap, there are two semantically tied but separate issues:
1. Controller bean preserved when navigate to B, which is a bug.
2. To provide a capabilities to make injected bean be in scope of their owner component, so as to have the same instance of beans within this owner component and it's child components hierarchy, and have another bean instance within another owner component hierarchy. This is a a new feature and should be extracted to a new issue.

**Comment by mshabarov** [comment]
Here is an example how to share the same instance of bean within some route/view and it's subcomponents, and to have a new bean instance upon navigating to another route/view:

```java
@UIScoped
@Route("institution")
public class EducationInstitutionView extends Div implements RouterLayout {

    private String uid = UUID.randomUUID().toString();

    public EducationInstitutionView() {
        add(new Span("This is an institution dashboard " + uid));
    }

    @Route(value = "elementary-school", layout = EducationInstitutionView.class)
    static class ElementarySchoolView extends Div {
        @Inject
        public ElementarySchoolView(@RouteScopeOwner(ElementarySchoolView.class)
                                            InstitutionService institutionService,
                                    SchoolStudentsCountComponent countView,
                                    SchoolStudentsListComponent listView) {
            add(new VerticalLayout(
                    new Span("This is " + getClass().getSimpleName()),
                    new Span("InstitutionService instance = " + institutionService.getUid()),
                    countView,
                    listView,
                    new RouterLink("Navigate to UniversityView", UniversityView.class)
            ));
        }
    }

    @Route(value = "university", layout = EducationInstitutionView.class)
    static class UniversityView extends Div {
        @Inject
        public UniversityView(@RouteScopeOwner(UniversityView.class)
                                      InstitutionService institutionService,
                              UniversityStudentsCountComponent countView,
                              UniversityStudentsListComponent listView) {
            add(new VerticalLayout(
                    new Span("This is " + getClass().getSimpleName()),
                    new Span("InstitutionService instance = " + institutionService.getUid()),
                    countView,
                    listView,
                    new RouterLink("Navigate to ElementarySchoolView", ElementarySchoolView.class)
            ));
        }
    }

    static class StudentsCountComponent extends Div {
        public StudentsCountComponent(InstitutionService institutionService) {
            add(new VerticalLayout(
                    new Span("This is " + getClass().getSimpleName()),
                    new Span("Students count = " + institutionService.getStudentsCount()),
                    new Span("InstitutionService instance = " + institutionService.getUid())));
        }
    }

    static class SchoolStudentsCountComponent extends StudentsCountComponent {
        @Inject
        public SchoolStudentsCountComponent(
                @RouteScopeOwner(ElementarySchoolView.class)
                        InstitutionService institutionService) {
            super(institutionService);
        }

        // School student concrete implementation
    }

    static class UniversityStudentsCountComponent extends StudentsCountComponent {
        @Inject
        public UniversityStudentsCountComponent(
                @RouteScopeOwner(UniversityView.class)
                        InstitutionService institutionService) {
            super(institutionService);
        }

        // University student concrete implementation
    }

    static class StudentsListComponent extends Div {
        public StudentsListComponent(InstitutionService institutionService) {
            add(new VerticalLayout(
                    new Span("This is " + getClass().getSimpleName()),
                    new Span("Students list = " + institutionService.getStudents()),
                    new Span("InstitutionService instance = " + institutionService.getUid())));
        }
    }

    static class SchoolStudentsListComponent extends StudentsListComponent {
        @Inject
        public SchoolStudentsListComponent(
                @RouteScopeOwner(ElementarySchoolView.class) InstitutionService institutionService) {
            super(institutionService);
        }

        // School student concrete implementation
    }

    static class UniversityStudentsListComponent extends StudentsListComponent {
        @Inject
        public UniversityStudentsListComponent(
                @RouteScopeOwner(UniversityView.class) InstitutionService institutionService) {
            super(institutionService);
        }

        // University student concrete implementation
    }

    @NormalRouteScoped
    @RouteScopeOwner(ElementarySchoolView.class)
    static class ElementarySchoolService extends InstitutionService {

        @Override
        public int getStudentsCount() {
            return 3;
        }

        @Override
        public Collection<String> getStudents() {
            return IntStream.range(1, 4)
                    .mapToObj(i -> "School Student-" + i)
                    .collect(Collectors.toList());
        }
    }

    @NormalRouteScoped
    @RouteScopeOwner(UniversityView.class)
    static class UniversityService extends InstitutionService {
        @Override
        public int getStudentsCount() {
            return 2;
        }

        @Override
        public Collection<String> getStudents() {
            return IntStream.range(1, 3)
                    .mapToObj(i -> "University Student-" + i)
                    .collect(Collectors.toList());
        }
    }
    
    static abstract class InstitutionService {
        private String uid = UUID.randomUUID().toString();
        public abstract int getStudentsCount();
        public abstract Collection<String> getStudents();
        public String getUid() { return uid; }
    }
}
```

Please note that I intentionally use two different implementations for `InstitutionService`, because It feels to me that when you have two different _views_ probably with the similar component structure and common behaviour, you still need two different implementations of service.

@thigg 

Do you have in mind any example when you want to share the same implementation (but different instances) between two different views? (same impl of `Controller` for `A` and `B` from your example). It feels for me that it's not a proper design. You'd rather share it, or use different implementations because the views are supposed to be different, so beans are also provide data in different ways.

Do you want to share the bean/service instance always within only the closest `@Route`/`RouterLayout` parent owner (and it's subcomponents), or you'd like to have an opportunity to choose it by yourself (for instance, make it being shared within some hierarchy of `@Route`/`RouterLayout` components)?

P.S. my previous example is not correct, because it injects a bean inside route component, which is out-of-scope of this bean, so it doesn't make sense semantically:

```java
@Inject
 public B(@RouteScopeOwner(A.class) GUIDController controller) {...}
```

**Comment by thigg** [comment]
Ah I'm sorry, I can't even remember what originally brung up the issue, as this is already almost a year ago.

We migrated a large piece of Vaadin8-Software and refactored quite a lot and in the end we solved the problem differently.
Might be that this was a code smell anyways and you actually would want two different implementations of the service.

If I come up with any example, I'll comment here.

**Comment by mshabarov** [comment]
@thigg thank you for feedback! 
Sure, if you come up with a use case, please share it with us. It would help us to advance in improving scopes behaviour in CDI and Spring add-ons.

At the moment, I believe the current capabilities of `@RouteScoped` and `@NormalRouteScoped` covers your case with A and B sub-hierarchies. The example above shows the usage of the bean/service.

Concerning sub-component injection I would propose either to **use different components (i.e. implementations)** and set their parent owner as a A or B, similarly like I did with service beans, or **extend the sub-component life span to parent layout scope or UI scope**, so as the same instance would be reused within A and B hierarchy.
Using the same instance of component within A hierarchy and another instance within B hierarchy feels a bit artificial, and I couldn't even suppose where it might be used.

So I'm gonna close this issue for now.
Anyway, please comment on this issue or open a new one, if you get more ideas or use cases.
